### ISSUE TITLE
🤖 Code Audit: 19 potential issue(s) found

### ISSUE THREAD

**Issue by asmit25805** [apple/coreai-models#44]
##  Code Audit Report

> All findings are reviewed for confidence before posting.
> Please verify each finding before acting on it.

**Repository:** `apple/coreai-models`
**Findings:** 19 issue(s) found — 🔴 1 critical · 🟠 13 high · 🟡 4 medium · 🔵 1 low

---

### 1. 🐛 Using torch.autocast on CPU with float16 dtype

| Field | Details |
|---|---|
| **Severity** | 🔴 Critical |
| **Type** | Bug |
| **File** | `models/clap/export.py` |
| **Location** | create_clap |
| **Confidence** | 92% |

**Problem:**
The code enables `torch.autocast(device_type="cpu", dtype=dtype)` regardless of the chosen dtype. CPU autocast does not support `float16`; attempting to run with `dtype=torch.float16` will raise a runtime error, preventing model export for half‑precision.

**Suggested Fix:**
Guard the autocast block to run only when `dtype` is a supported CPU dtype (e.g., `torch.float32`) or remove the autocast entirely for CPU execution. Example:
```python
if dtype == torch.float32:
    with torch.autocast(device_type="cpu", dtype=dtype):
        exported = torch.export.export(...)
else:
    exported = torch.export.export(...)
```

---

### 2. 🐛 Incorrect tensor dtype for input IDs and attention mask

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Bug |
| **File** | `models/clap/export.py` |
| **Location** | reference_inputs |
| **Confidence** | 96% |

**Problem:**
The function `reference_inputs` casts `input_ids` and `attention_mask` to `torch.int32`. HuggingFace models expect these tensors to be of type `torch.long` (int64). Supplying int32 tensors can cause shape or type errors during model execution, leading to crashes or incorrect results.

**Suggested Fix:**
Replace `.to(torch.int32)` with `.to(torch.long)` (or omit the cast, as the processor already returns the correct dtype).

---

### 3. 🐛 Incorrect tensor dtype for token IDs

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Bug |
| **File** | `models/clip/export.py` |
| **Location** | reference_inputs |
| **Confidence** | 95% |

**Problem:**
The function `reference_inputs` casts `input_ids` and `attention_mask` to `torch.int32`. The CLIP model expects these tensors to be of type `torch.long` (int64). Supplying int32 tensors can cause type mismatches during export or inference, leading to runtime errors or incorrect results.

**Suggested Fix:**
Remove the explicit `.to(torch.int32)` casts and let the tokenizer produce tensors of the default `torch.long` type, or explicitly cast to `torch.long` using `.to(torch.long)`.

---

### 4. 🐛 torch.autocast on CPU does not support float16

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Bug |
| **File** | `models/efficient-sam/export.py` |
| **Location** | with torch.autocast(device_type="cpu", dtype=dtype): |
| **Confidence** | 95% |

**Problem:**
The code unconditionally uses torch.autocast with device_type="cpu" and the user‑selected dtype. PyTorch's CPU autocast only supports bfloat16; using float16 will raise a RuntimeError. The earlier check only blocks float16 when dynamic is enabled, leaving a path where static float16 is still used and will crash at runtime.

**Suggested Fix:**
Add a guard to disable autocast when dtype is torch.float16 on CPU, e.g., skip the autocast block or switch to a supported dtype. Alternatively, move the export to GPU if float16 is required.

---

### 5. 🐛 Invalid autocast dtype for CPU device

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Bug |
| **File** | `models/pvt/export.py` |
| **Location** | create_pvt -> torch.autocast(device_type="cpu", dtype=dtype) |
| **Confidence** | 95% |

**Problem:**
The code uses torch.autocast with device_type="cpu" and allows float16 (torch.float16) as the dtype. PyTorch's CPU autocast only supports bfloat16; using float16 will raise a RuntimeError, causing the script to fail for the float16 option.

**Suggested Fix:**
Restrict the allowed dtypes for CPU autocast to bfloat16 only, or move the autocast to a GPU device when float16 is selected. For example, change the dtype mapping to only include bfloat16 for CPU, or add a conditional: if dtype == torch.float16: raise an informative error or skip autocast.

---

### 6. 🐛 Incorrect dtype for input_ids tensor

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Bug |
| **File** | `models/roberta/export.py` |
| **Location** | reference_inputs function (input_ids conversion) |
| **Confidence** | 95% |

**Problem:**
The code converts the token IDs returned by the tokenizer from the default torch.int64 (torch.long) to torch.int32. RoBERTa's embedding layer expects 64‑bit integer indices; using int32 will cause a type mismatch error during export or inference, leading to a runtime failure.

**Suggested Fix:**
Remove the .to(torch.int32) conversion or explicitly cast to torch.long (torch.int64). For example: `inputs = {"input_ids": input_ids}` or `inputs = {"input_ids": input_ids.to(torch.long)}`. Ensure any subsequent operations (e.g., expand) preserve the long dtype.

---

### 7. 🐛 Incorrect dtype for token IDs

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Bug |
| **File** | `models/sam3/export.py` |
| **Location** | reference_inputs function |
| **Confidence** | 92% |

**Problem:**
The function converts the tokenizer's `input_ids` tensor to `torch.int32`. Token IDs are typically 64‑bit integers (`torch.int64`). Down‑casting to 32‑bit can cause overflow for vocabularies larger than 2^31 and may break downstream model expectations, leading to runtime errors or incorrect predictions.

**Suggested Fix:**
Remove the explicit cast or change it to `torch.int64` (the default). For example: `input_ids = text_inputs["input_ids"]` without `.to(torch.int32)`.

---

### 8. 🐛 Incorrect dtype for token IDs

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Bug |
| **File** | `models/t5/export.py` |
| **Location** | reference_inputs |
| **Confidence** | 96% |

**Problem:**
The function `reference_inputs` casts `input_ids` and `decoder_input_ids` to `torch.int32`. HuggingFace transformer models expect token IDs as `torch.long` (int64). Supplying int32 tensors can cause type mismatches or runtime errors during model export or inference.

**Suggested Fix:**
Remove the `.to(torch.int32)` casts or replace them with `.to(torch.long)` to keep the default dtype expected by the model.

---

### 9. 🐛 Unsupported float16 autocast on CPU

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Bug |
| **File** | `models/t5/export.py` |
| **Location** | create_t5 |
| **Confidence** | 94% |

**Problem:**
The code uses `torch.autocast(device_type="cpu", dtype=dtype)` where `dtype` may be `torch.float16`. PyTorch's autocast on CPU only supports `bfloat16`; using `float16` will raise a runtime error, preventing the export from succeeding.

**Suggested Fix:**
Either restrict the `--dtype` option to `float32` for CPU execution, or change the autocast device to `cuda` when `float16` is selected. Alternatively, use `torch.bfloat16` for CPU autocasting.

---

### 10. 🐛 torch.autocast on CPU with float16 may raise runtime error

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Bug |
| **File** | `models/wav2vec2/export.py` |
| **Location** | create_wav2vec2 -> torch.autocast |
| **Confidence** | 92% |

**Problem:**
The code unconditionally wraps the export call in `torch.autocast(device_type="cpu", dtype=dtype)`. CPU autocast only supports bfloat16; using `torch.float16` (or other unsupported dtypes) will cause a runtime error, breaking the export process for the selected dtype. This is a functional bug that prevents successful model conversion when the user selects float16.

**Suggested Fix:**
Guard the autocast usage: only enable it when the device and dtype are supported (e.g., if torch.cuda.is_available() and dtype is float16, use device_type="cuda"; otherwise skip autocast or use bfloat16 for CPU). Example:
```python
if torch.cuda.is_available() and dtype == torch.float16:
    ctx = torch.autocast(device_type="cuda", dtype=dtype)
else:
    ctx = contextlib.nullcontext()
with ctx:
    exported = torch.export.export(...)
```

---

### 11. 🐛 Incomplete conditional block causing syntax error

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Bug |
| **File** | `python/src/coreai_models/model_registry.py` |
| **Location** | filter_presets function (around line where 'if not include_experimental and p.experimental:' is) |
| **Confidence** | 99% |

**Problem:**
The 'filter_presets' function contains an 'if' statement without a body. This results in a SyntaxError / IndentationError, preventing the module from being imported or executed. The function is intended to filter presets based on criteria, but the missing implementation aborts the code path.

**Suggested Fix:**
Complete the conditional block by adding the appropriate logic (e.g., 'continue' to skip experimental presets) and implement the remaining filtering criteria. Ensure the function ends with 'return out'. Example:
```python
def filter_presets(
    presets: Iterable[ModelPreset],
    *,
    family: str | None = None,
    variant: str | None = None,
    include_experimental: bool = False,
) -> list[ModelPreset]:
    out: list[ModelPreset] = []
    for p in presets:
        if not include_experimental and p.experimental:
            continue
        if family is not None and p.family != family:
            continue
        if variant is not None and p.variant != variant:
            continue
        out.append(p)
    return out
```

---

### 12. 🐛 Potential out‑of‑bounds indexing when constructing dummy RoPE IDs

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Bug |
| **File** | `python/src/coreai_models/diffusion/flux2.py` |
| **Location** | _dummy_flux2_transformer_impl (img_ids/ txt_ids indexing) |
| **Confidence** | 96% |

**Problem:**
The dummy input factory assumes that the RoPE axes dimension list has at least three (for img_ids) and four (for txt_ids) entries and writes to hard‑coded indices 1, 2, and 3. If the model configuration provides fewer axes, this will raise an IndexError, breaking dummy‑input generation and any export pipelines that rely on it.

**Suggested Fix:**
Replace the hard‑coded index assignments with a loop that respects the actual number of axes, e.g., fill the first few axes with position values up to min(num_rope_axes, required_axes).

---

### 13. 🐛 Incorrect attribute access on VAE encoder output

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Bug |
| **File** | `python/src/coreai_models/diffusion/flux2.py` |
| **Location** | Flux2VAEEncoderWrapper.forward |
| **Confidence** | 94% |

**Problem:**
The forward method returns `self.vae.encode(x).latent_dist.parameters`, but `latent_dist` is a distribution object, not a nn.Module. The `parameters` attribute does not exist, causing an AttributeError at runtime and preventing the encoder wrapper from producing latent tensors.

**Suggested Fix:**
Return the actual latent tensor, e.g., `self.vae.encode(x).latent_dist.sample()` or `self.vae.encode(x).latent_dist.mean`, depending on the intended behavior.

---

### 14. 🐛 Potential mismatch between dummy_inputs and input/output name tuples

| Field | Details |
|---|---|
| **Severity** | 🟠 High |
| **Type** | Bug |
| **File** | `python/src/coreai_models/diffusion/gpu.py` |
| **Location** | export_stateless function |
| **Confidence** | 92% |

**Problem:**
The function passes dummy_inputs to torch.export.export and later provides input_names and output_names to coreai_torch.TorchConverter.add_pytorch_module without verifying that the number of tensors in dummy_inputs matches the length of input_names (and similarly for output_names). A mismatch will cause runtime errors during export or conversion.

**Suggested Fix:**
Add explicit validation at the start of export_stateless to ensure len(dummy_inputs) == len(input_names) and that the number of outputs produced by the model matches len(output_names). Raise a clear exception if they differ.

---

### 15. 🔒 Potential unsafe deletion of user‑specified output directory

| Field | Details |
|---|---|
| **Severity** | 🟡 Medium |
| **Type** | Security |
| **File** | `models/clip/export.py` |
| **Location** | _save_asset |
| **Confidence** | 85% |

**Problem:**
When `overwrite` is true and the target `model_path` already exists, the code deletes the existing path. If `output_dir` is supplied by an untrusted user, a crafted path could cause `shutil.rmtree` to remove arbitrary directories (e.g., root or system folders) if they match the constructed `model_path`. This presents a directory‑traversal / deletion risk.

**Suggested Fix:**
Validate `output_dir` to ensure it is within an allowed directory tree, reject absolute paths outside the intended location, and only delete the specific file rather than an entire directory. Consider using `model_path.is_file()` check before removal and avoid `shutil.rmtree` on directories.

---

### 16. 🔒 Downloading model weights from an external URL without verification

| Field | Details |
|---|---|
| **Severity** | 🟡 Medium |
| **Type** | Security |
| **File** | `models/efficient-sam/export.py` |
| **Location** | torch.hub.load_state_dict_from_url( |
| **Confidence** | 88% |

**Problem:**
The script fetches the EfficientSAM checkpoint directly from a HuggingFace URL using torch.hub.load_state_dict_from_url, which does not verify the file's integrity (no checksum) and trusts the remote source. This opens a supply‑chain attack vector where a malicious actor could replace the model file and execute arbitrary code when the state dict is loaded.

**Suggested Fix:**
Host the checkpoint on a trusted internal server or include a known SHA256 checksum and verify it after download before loading. Use torch.hub.load_state_dict_from_url with the 'hash' parameter or manually download and compare hashes.

---

### 17. 🐛 CPU autocast with float16 may raise runtime error

| Field | Details |
|---|---|
| **Severity** | 🟡 Medium |
| **Type** | Bug |
| **File** | `models/sam3/export.py` |
| **Location** | create_sam3 function (torch.autocast block) |
| **Confidence** | 85% |

**Problem:**
The code uses `torch.autocast(device_type="cpu", dtype=dtype)` where `dtype` can be `torch.float16`. CPU autocast currently does not support float16 and will raise an exception when `dtype` is float16, preventing the export from succeeding.

**Suggested Fix:**
Either restrict the allowed dtypes to those supported on CPU (e.g., only `torch.float32`) or switch to a GPU device when `dtype` is float16. Example: `device = "cuda" if torch.cuda.is_available() else "cpu"; with torch.autocast(device_type=device, dtype=dtype):`.

---

### 18. 💡 In-place modification of the model by calling eval()

| Field | Details |
|---|---|
| **Severity** | 🟡 Medium |
| **Type** | Suggestion |
| **File** | `python/src/coreai_models/diffusion/gpu.py` |
| **Location** | wrapper.eval() call |
| **Confidence** | 85% |

**Problem:**
Calling wrapper.eval() changes the module's training mode globally, which may be unexpected for callers that reuse the same wrapper elsewhere. This side‑effect can lead to subtle bugs if the model is later used for training.

**Suggested Fix:**
Replace the in‑place eval() call with a context manager that temporarily sets the module to eval mode, e.g., using torch.no_grad() and wrapper.train(False) within a try/finally block, or document the side‑effect clearly.

---

### 19. 💡 Avoid recreating constant tensors on each call

| Field | Details |
|---|---|
| **Severity** | 🔵 Low |
| **Type** | Suggestion |
| **File** | `models/efficient-sam/export.py` |
| **Location** | batched_points = torch.rand((B, num_queries, num_pts, 2)) * torch.tensor([H, W]) |
| **Confidence** | 92% |

**Problem:**
In reference_inputs, a new 1‑D tensor [H, W] is allocated on every invocation, which is unnecessary overhead for a constant value. While not a functional bug, it adds minor CPU work and memory churn during repeated calls.

**Suggested Fix:**
Define a module‑level constant, e.g., _HW = torch.tensor([1024, 1024], dtype=torch.float32), and reuse it, or multiply by a scalar tuple (H, W) directly.

---

<details>
<summary>About this report</summary>

This report was generated using Llama 3.3 70B.
Only findings with ≥80% confidence are included.
False positives are possible — use your own judgment.

</details>

**Comment by adityabagchi24** [comment]
Hi! I investigated the issue "Using torch.autocast on CPU with float16 dtype" and implemented a fix in my fork.

The change avoids entering the CPU autocast context when dtype=torch.float16, since CPU autocast does not support float16 and raises a runtime error during export.

Commit: https://github.com/adityabagchi24/coreai-models/commit/b39d55df51267ee735052485f2d8af23f1702291

If this approach looks reasonable, feel free to cherry-pick the commit or let me know if you'd prefer a different implementation.

**Comment by stikves** [comment]
Thank you for the report, but many of these seem to miss out on the actual design choices, or lower prioirty

Please send individual requests after verifying them in practice.

For example:
* 2, 3, 6, 7, 8: Core AI uses `int32`, not `int64`, hence those casts are valid.

Other only trigger on non-standard paths.

**Comment by stikves** [comment]
@adityabagchi24 thank you

However that path would only trigger if the user specifically asks for `--dtype float16` for a CPU model. We might later add a check in the command line flag parser to warn them earlier.

**Comment by asmit25805** [comment]
Thanks @stikves and @adityabagchi24  for reviewing this report, but this report was generated during the testing phase of my system so false positives may be possible. The system is continuously being updated and improved, and any recommendations you have would be highly appreciated.
we can connect on IG @buildwithasmit for some discussions and meaningful conversations.
Thank you,
Asmit

**Comment by stikves** [comment]
Thank you for your interest and hopefully these models would be useful in your projects

However, automated tools bring many false positives. If you have control over them, it might help asking to verify their assertions locally, or have them look at larger context.
